contracts/interchain-token/src/utils.rs

```diff
@@ -0,0 +1,145 @@
+use soroban_sdk::{Address, Env, String};
+
+use crate::storage_types::{
+    AllowanceDataKey, AllowanceValue, DataKey, BALANCE_BUMP_AMOUNT, BALANCE_LIFETIME_THRESHOLD,
+    INSTANCE_BUMP_AMOUNT, INSTANCE_LIFETIME_THRESHOLD,
+};
+use soroban_token_sdk::metadata::TokenMetadata;
+use soroban_token_sdk::TokenUtils;
+
+pub fn check_nonnegative_amount(amount: i128) {
+    if amount < 0 {
+        panic!("negative amount is not allowed: {}", amount)
+    }
+}
+
+pub fn extend_instance_ttl(env: &Env) {
+    env.storage()
+        .instance()
+        .extend_ttl(INSTANCE_LIFETIME_THRESHOLD, INSTANCE_BUMP_AMOUNT);
+}
+
+pub fn extend_balance_ttl(env: &Env, key: &DataKey) {
+    env.storage()
+        .persistent()
+        .extend_ttl(key, BALANCE_LIFETIME_THRESHOLD, BALANCE_BUMP_AMOUNT);
+}
+
+pub fn admin(env: &Env) -> Address {
+    env.storage().instance().get(&DataKey::Admin).unwrap()
+}
+
+pub fn read_allowance(env: &Env, from: Address, spender: Address) -> AllowanceValue {
+    let key = DataKey::Allowance(AllowanceDataKey { from, spender });
+    env.storage()
+        .temporary()
+        .get::<_, AllowanceValue>(&key)
+        .map_or(
+            AllowanceValue {
+                amount: 0,
+                expiration_ledger: 0,
+            },
+            |allowance| {
+                if allowance.expiration_ledger < env.ledger().sequence() {
+                    AllowanceValue {
+                        amount: 0,
+                        expiration_ledger: allowance.expiration_ledger,
+                    }
+                } else {
+                    allowance
+                }
+            },
+        )
+}
+
+pub fn write_allowance(
+    env: &Env,
+    from: Address,
+    spender: Address,
+    amount: i128,
+    expiration_ledger: u32,
+) {
+    let allowance = AllowanceValue {
+        amount,
+        expiration_ledger,
+    };
+
+    if amount > 0 && expiration_ledger < env.ledger().sequence() {
+        panic!("expiration_ledger is less than ledger seq when amount > 0")
+    }
+
+    let key = DataKey::Allowance(AllowanceDataKey { from, spender });
+    env.storage().temporary().set(&key, &allowance);
+
+    if amount > 0 {
+        let live_for = expiration_ledger
+            .checked_sub(env.ledger().sequence())
+            .unwrap();
+
+        env.storage()
+            .temporary()
+            .extend_ttl(&key, live_for, live_for)
+    }
+}
+
+pub fn spend_allowance(env: &Env, from: Address, spender: Address, amount: i128) {
+    let allowance = read_allowance(env, from.clone(), spender.clone());
+    if allowance.amount < amount {
+        panic!("insufficient allowance");
```

Comment:
use ensure!

Reply:
create an `assert_with_error!` macro that uses `panic_with_error!` underthehood maybe? alternatively return a Result from here, and then use `.expect()` in the caller